### PR TITLE
WIP: Lookup fonts from filesystem

### PR DESCRIPTION
**Pull request by sp1ritCS**
In #87 you've said that
>> fallback fonts
>
> Probably no. In my submission, it is the user's responsibility to provide the font with the clm data file, but it needs to be discussed.

But for [NoteKit](https://github.com/blackhole89/notekit), we'd prefer if font handling will be done by clatexmath, rather than notekit. This PR is basically implementing a way of doing this:

It first looks at all potential tinytex data dirs (`$TINYTEX_FONTDIR`, `$XDG_DATA_HOME/tinytex`, `$XDG_DATA_DIRS[@]/tinytex`, `/usr/local/share/tinytex` and `/usr/share/tinytex`) and in for there to be a valid font there needs to be a `FontName.otf` and a `FontName.clm` file in these directories. 

The clm file spec has been changed to include postscript full name and postscript family name after the `has_glyph_path` byte (see the BinaryReader::read_string and lookfwd method), so that the found clm files can be parsed and included in the FontContext as their name.

To accommodate for this, the `LaTeX::init` method had to be changed. Instead of a `const FontSrc&` it now takes a `std::variant<const FontSrc*, const std::string*, {}>`, to either: init clm with a fontspec, like it worked before (alsotho variants can't seem to contain references, so one now needs to borrow the fontspec when initializing), init clm with a specific mathfont that the progrom will expect to be found or with nothing (`InitFontSenseAuto`), then first mathfont that is encountered will also become the default math font.

Currently I'll label this PR as a WIP, because there are still a few bugs to fix and it can surely be still improved.

I also found a bug (I think), in the way non math font families are handled. I've tried to put NoteKit's fontfamily "Charter" (https://github.com/blackhole89/notekit/tree/master/data/fonts) into the search dir, and the font senser also found all of them and added all of them with a single `LaTeX::addMainFont` invocation as "Charter" family. However, after calling `LaTeX::setDefaultMainFont("Charter")` instead of picking the correct style for `\text{}` it instead used the one that was in the last position of the FontSrcList, which happened to be "Charter Italic".

---

## TODO:
- [ ] fix the bug described above
- [x] windows compatibly. I'm unable to test against any Windows OS, so I don't know if it works there as the windows CI seems currently broken. Additionally, I'd like to extend the font senser for windows to look for $DIR_OF_EXECUTABLE/share/tinytex, so our notekit windows build will continue to work. I imagine this similar to glib's [g_win32_get_system_data_dirs_for_module](https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gutils.c#L2527) which is used in [g_get_system_data_dirs](https://docs.gtk.org/glib/func.get_system_data_dirs.html) on windows. maybe @PikachuHy could help here.

### CODE REVIEW

File: lib/otf/fontsense.cpp
Comment (NanoMichael) [inline lib/otf/fontsense.cpp:18]: This project prefers 2 spaces indention, recommend following clang-format-style:

```
{
    BasedOnStyle: Google,
    IndentWidth: 2,
    AccessModifierOffset: -2,
    AllowShortBlocksOnASingleLine: true,
    AllowShortIfStatementsOnASingleLine: true,
    AllowShortCaseLabelsOnASingleLine: true,
    AllowShortLoopsOnASingleLine: true,
    ColumnLimit: 0,
    AlignAfterOpenBracket: false,
    BinPackArguments: false,
    BinPackParameters: true,
    AllowAllParametersOfDeclarationOnNextLine: true,
    ReflowComments: true,
    AlignConsecutiveMacros: false,
    AlignTrailingComments: true,
    AlignConsecutiveAssignments: false,
    AlignConsecutiveDeclarations: false,
    ContinuationIndentWidth: 2
}
```

I'll make a contribution guide ASAP.

File: lib/tinytex.cpp
Comment (NanoMichael) [inline lib/tinytex.cpp:21]: I believe directly using `if (_isInited) return;` is better.

File: lib/tinytex.cpp
Comment (NanoMichael) [inline lib/tinytex.cpp:18]: using `if(mathFontSrc != nullptr)` is more clear.

File: lib/otf/fontsense.h
Comment (NanoMichael) [inline lib/otf/fontsense.h:15]: This project prefers camel case naming styles.

File: lib/tinytex.cpp
Comment (sp1ritCS) [reply lib/tinytex.cpp:21]: the initialization is not just `if (_isInited) return;`, but also includes
```cpp
  initialization:if (_isInited) return;
  _isInited = true;
  NewCommandMacro::_init_();
```

**Comment by NanoMichael** [comment]
> I'm unable to test against any Windows OS

I got a Windows PC :smirk: I'll check this

**Comment by sp1ritCS** [comment]
>> I'm unable to test against any Windows OS

> I got a Windows PC 😏 I'll check this

well, now the CI works again, I fixed the issues with MSVCs standard library (unlike GNUs libstd++, std::filesystem::path doesn't have a c_str method directly).

The only other real issue I could imagine happening is are issues with reading the UTF-8 strings from the clm files, given that windows usually uses utf-16.

File: lib/tinytex.cpp
Comment (NanoMichael) [reply lib/tinytex.cpp:21]: I've missed these lines... maybe lambda is a better choice?

```C++
auto initialization = [&]() {
  if (_isInited) return;
  _isInited = true;
  NewCommandMacro::_init_();
}
```

**Comment by NanoMichael** [comment]
> The only other real issue I could imagine happening is are issues with reading the UTF-8 strings from the clm files, given that windows usually uses utf-16.

I believe it should not be a problem. `Otf::name()` and `Otf::family()` returns utf-8 byte sequence, the Windows user can convert it to utf-16 string if they want.

File: lib/otf/fontsense.cpp
Comment (PikachuHy) [reply lib/otf/fontsense.cpp]: We can add a .clang-format file to the project.

**Comment by PikachuHy** [comment]
> I'm unable to test against any Windows OS

I will check it later.

File: prebuilt/otf2clm.py
Comment (sp1ritCS) [reply prebuilt/otf2clm.py:646]: sure, I just assumed that doesn't play a role yet, only after unimath gets officially released

Comment (NanoMichael) [reply prebuilt/otf2clm.py]: yeah, just in case someone uses it mistakenly

**Comment by PikachuHy** [comment]
> Additionally, I'd like to extend the font senser for windows to look for $DIR_OF_EXECUTABLE/share/tinytex

is this feature implemented?

**Comment by sp1ritCS** [comment]
@PikachuHy 
> is this feature implemented?

no, I asked If you could help with that. Take a look at the TODO section at the bottom of the PR.

So far on windows it'll just check `$TINYTEX_FONTDIR` and `$XDG_DATA_DIRS`/`$XDG_DATA_HOME` for mingw envs.

**Comment by PikachuHy** [comment]
why the font not work automatically?

prepare font
image
I set `TINYTEX_FONTDIR` to the font path

test code
```cpp
using namespace tinytex;
int main() {
  InitFontSenseAuto init;
  try {
    TinyTeX::init(init);
  } catch (std::exception& e) {
    std::cout << "exception msg: " << e.what() << std::endl;
  }
  return 0;
}
```

run log
```
C:\code\cLaTeXMath\cmake-build-debug\tinytex-qt-example.exe
family: XITS
found: XITS
exception msg: no math font found by fontsense
```

**Comment by sp1ritCS** [comment]
@PikachuHy, XITS is not the Math font, its a Regular one. XITS Math or LatinModern Math are math fonts.

**Comment by sp1ritCS** [comment]
and it needs at least one mathfont in order to work, regular fonts are optional.

**Comment by PikachuHy** [comment]
how to test on notekit?

**Comment by sp1ritCS** [comment]
notekit doesn't have support for tinytex/unimath yet.

we wait until it's merged into master

**Comment by sp1ritCS** [comment]
just the one about that .clang format file. nothing else

**Comment by PikachuHy** [comment]
> just the one about that .clang format file. nothing else

OK. Something may have gone wrong. I resend my comments.

1.

https://github.com/sp1ritCS/cLaTeXMath/blob/3ed6fcd06924cb8484e486f10e4735a1eb154cc7/lib/otf/fontsense.cpp#L71

suggestion: add a guard
It might be better to only consider xxx.otf, then check xxx.clm file exists.
for example
```
if (ext == "otf" and stem.clm file exist) {
    // add to font_paths
} else {
    // ignore
}

```

2.
https://github.com/sp1ritCS/cLaTeXMath/blob/3ed6fcd06924cb8484e486f10e4735a1eb154cc7/lib/tinytex.cpp#L36

indent error
use space instead of tab

3.
https://github.com/sp1ritCS/cLaTeXMath/blob/3ed6fcd06924cb8484e486f10e4735a1eb154cc7/lib/otf/fontsense.cpp#L50


> Additionally, I'd like to extend the font senser for windows to look for $DIR_OF_EXECUTABLE/share/tinytex, so our notekit windows build will continue to work.

change code to
```c++
#else
  paths.push("share/" FONTDIR);
#endif
```
it works with such file tree
```
PS C:\code\cLaTeXMath\cmake-build-debug> .\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS C:\code\cLaTeXMath\cmake-build-debug> tree /f    
Folder PATH listing
Volume serial number is 6659-D38D
C:.
│  tinytex-qt-example.exe
│  tinytex-qt-example.ilk
│  tinytex-qt-example.pdb
│  tinytex-qt.dll
│  tinytex.dll
│  tinytex.ilk
│  tinytex.pdb
│
└─share
    └─tinytex
            XITSMath-Regular.clm
            XITSMath-Regular.otf
```

by the way, why are my comments not visible?
image

@NanoMichael Can you see my code review comments?

**Comment by NanoMichael** [comment]
> by the way, why are my comments not visible?

Your comment is now in "pending" state, you need to edit it via the "..." menu.

**Review by PikachuHy — CHANGES_REQUESTED** [review]

File: lib/otf/fontsense.cpp
Comment (sp1ritCS) [reply lib/otf/fontsense.cpp:50]: the problem is, that this will only work when the cwd is also the one in which the executable is located. what I'd prefer is that you could run the executable from anywhere (i.e. cwd is `Z:\home\florian\` but the binary is located in `C:\Program Files\Application\binary.exe`. To do that you need that function winapi has to figure out where the binary is located (similar to `dirname $(readlink("/proc/self/exe")` on POSIX) that glib uses too when getting the data dirs on windows.

File: lib/otf/fontsense.cpp
Comment (sp1ritCS) [reply lib/otf/fontsense.cpp:70]: I already filter out any elements where either path is NULL.

https://github.com/sp1ritCS/cLaTeXMath/blob/3ed6fcd06924cb8484e486f10e4735a1eb154cc7/lib/otf/fontsense.cpp#L92

this way the clm and the otf file could be in different search dirs, but still work in the end.

**Comment by sp1ritCS** [comment]
@PikachuHy 
> how to test on notekit?

I've actually tried porting notekit to use unimath right now, and I've had a few issues.
- is TINYTEX_EXPORT actually required in headerfiles? Because it isn't defined in notekit, the compiler considers TINYTEX_EXPORT the class name and then fails.
- the c cairo interface has these c ptr wrapper classes with destructors freeing the pointers. however there are functions defined in the headerfile that end up causing linker issues. @NanoMichael I'll submit a PR fixing that. Should I just inline them or put the definition the the source file?

**Comment by NanoMichael** [comment]
> is TINYTEX_EXPORT actually required in headerfiles?

No, it is not required, the macro TINYTEX_EXPORT is just to tell the linker to export the public symbols of the library, you can just make an empty macro definition ( just like `#define TINYTEX_EXPORT`) in order to make it work (but that is a temporary solution, I'll make an all-in-one header file to include all the public header files.)

> Should I just inline them or put the definition the the source file?

Impl at source file is preferred.

**Comment by sp1ritCS** [comment]
nah, both of my TODO's aren't yet solved. Have you looked at the issues with the font families yet?

**Comment by sp1ritCS** [comment]
>>  Should I just inline them or put the definition the the source file?
>
> Impl at source file is preferred.


See #103

**Comment by sp1ritCS** [comment]
> how to test on notekit?

>  you can make an empty macro definition (`#define TINYTEX_EXPORT`) in order to make it work

with that it works 🎉 (obs. this is just a quick attempt, I'll wait for you to make it properly. (also support for loading the Charter font is also not there yet))
https://github.com/sp1ritCS/notekit/commit/ae83ca817a27d97b7b59407a8c4037ef318830b9

File: lib/otf/fontsense.cpp
Comment (PikachuHy) [reply lib/otf/fontsense.cpp:50]: new solution

```cpp
#ifdef _WIN32
#include <windows.h>
// Get dir of executable
// For example, the full path is c:\x\y\z\a.exe
// return c:\x\y\z\
//
std::string getDirOfExecutable() {
  char szBuff[MAX_PATH] = {0};
  // WIN API: GetModuleFileNameA
  // retrieves the path of the executable file of the current process
  // see https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea
  GetModuleFileNameA(nullptr,szBuff, MAX_PATH);
  std::string exePath(szBuff);
  auto pos = exePath.find_last_of('\\');
  // pos + 1 for adding last \ to path
  return exePath.substr(0, pos + 1);
}
#endif
```
and then

```cpp
#else
  auto exeDir = getDirOfExecutable();
  paths.push(exeDir + std::string("share/" FONTDIR));
#endif
```

it works with such file tree
```
PS C:\code\cLaTeXMath\cmake-build-debug> tree /f    
Folder PATH listing
Volume serial number is 6659-D38D
C:.
│  tinytex-qt-example.exe
│  tinytex-qt-example.ilk
│  tinytex-qt-example.pdb
│  tinytex-qt.dll
│  tinytex.dll
│  tinytex.ilk
│  tinytex.pdb
│
└─share
    └─tinytex
            XITSMath-Regular.clm
            XITSMath-Regular.otf
```
test case
```
PS C:\code\cLaTeXMath\cmake-build-debug> .\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS C:\code\cLaTeXMath> .\cmake-build-debug\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS C:\code> .\cLaTeXMath\cmake-build-debug\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS D:\> c:\code\cLaTeXMath\cmake-build-debug\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
```

**Comment by sp1ritCS** [comment]
@NanoMichael Given that the font family issue is a more general thing better addressed separately, I believe you may merge this now.

**Comment by NanoMichael** [comment]
> Given that the font family issue is a more general thing better addressed separately, I believe you may merge this now.

That's fair enough. I've made a new issue to track this problem, see #105

**Pull request merged**
